File: fronted/src/types/role.ts
// 角色相关的TypeScript类型定义

export interface Role {
  id: number;
  name: string;
  prompt: string;
  created_at: string;
  updated_at?: string;
}

// 角色创建/更新请求类型
export interface RoleRequest {
  name: string;
  prompt: string;
}

// 角色列表查询参数
export interface RoleListParams {
  keyword?: string;
  page?: number;
  page_size?: number;
  type?: string;
}

// 后端API响应的基础类型
export interface ApiResponse<T = any> {
  success: boolean;
  data: T;
  message?: string;
  error_code?: string;
}

// 角色列表响应数据类型
export interface RoleListResponse {
  roles: Role[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

// 前端组件内部使用的角色列表类型
export interface RoleListResult {
  items: Role[];
  total: number;
  page?: number;
  page_size?: number;
}

// API错误类型
export interface ApiError {
  success: false;
  error_code: string;
  message: string;
}

// 角色类型枚举（基于后端Schema）
export type RoleType = 'teacher' | 'student' | 'expert' | 'reviewer' | 'official' | 'lawyer' | 'other';

File: backend/app/schemas/role.py
from marshmallow import Schema, fields, validate, validates, ValidationError


class RoleSchema(Schema):
    """角色模式"""
    id = fields.Integer(dump_only=True)
    name = fields.String(required=True, validate=validate.Length(min=1, max=100))
    type = fields.String(required=True, validate=validate.OneOf([
        'teacher', 'student', 'expert', 'reviewer', 'official', 'lawyer', 'other'
    ]))
    description = fields.String(required=True, validate=validate.Length(min=1, max=1000))
    style = fields.String(validate=validate.Length(max=200))
    constraints = fields.String(validate=validate.Length(max=500))
    focus_points = fields.List(fields.String(), validate=validate.Length(max=10))
    is_builtin = fields.Boolean(dump_only=True)
    created_at = fields.DateTime(dump_only=True)
    updated_at = fields.DateTime(dump_only=True)

    @validates('name')
    def validate_name(self, value):
        """验证角色名称唯一性"""
        from app.models import Role
        if self.context.get('role_id'):  # 更新场景
            if Role.query.filter(Role.name == value, Role.id != self.context['role_id']).first():
                raise ValidationError('角色名称已存在')
        else:  # 创建场景
            if Role.query.filter_by(name=value).first():
                raise ValidationError('角色名称已存在')


class RoleListSchema(Schema):
    """角色列表模式"""
    roles = fields.List(fields.Nested(RoleSchema()))
    total = fields.Integer()
    page = fields.Integer()
    page_size = fields.Integer()